### PR TITLE
Improve update of datasets to use correct dataset order

### LINKED ISSUE #226: The order of series is not maintained when updating a chart
I have a chart with multiple series which are toggled outside the chart. When I toggle an element of an then on again it is placed at the end of the dataseries array instead of its intended location. This is a problem for me as it messes up the z-index and also the legend is incorrectly ordered.

I have already looked at the code and will create a PR for this issue.

### PR DESCRIPTION
**Pull request by sjurba**
This fixes #226 by maintaining the order of the nextDatasets while merging the data from currentDatasets.

Fixes #226

### CODE REVIEW

**Comment by benmccann** [comment]
@sjurba this PR will need to be rebased

**Comment by sjurba** [comment]
@benmccann: reabased onto master.

File: src/index.js [hunk 1]
```diff
@@ -160,36 +160,35 @@ class ChartComponent extends React.Component {
...
+      this.props.datasetKeyProvider
+    );
+
+    //We can safely replace the dataset array, as long as we retain the _meta property
+    //on each dataset.
+    this.chartInstance.config.data.datasets = nextDatasets.map(next => {
+      const current =
+        currentDatasetsIndexed[this.props.datasetKeyProvider(next)];
+      if (current && current.type === next.type) {
+        //The data array must be edited in place. As chart.js adds listeneres to it.
```
Comment (benmccann) [inline src/index.js]: listeneres -> listeners

File: src/index.js [hunk 2]
```diff
@@ -160,36 +160,35 @@ class ChartComponent extends React.Component {
...
+        current.data.splice(next.data.length);
+        next.data.forEach((point, pid) => {
+          current.data[pid] = next.data[pid];
+        });
+        const { data, ...otherProps } = next;
+        //Merge properties. Notice weekness here if a property is removed
+        //from next, it will be retained by current and never dissapears.
```
Comment (benmccann) [inline src/index.js]: dissapears -> disappears

File: src/index.js [hunk 3]
```diff
@@ -160,36 +160,35 @@ class ChartComponent extends React.Component {
...
+      const current =
+        currentDatasetsIndexed[this.props.datasetKeyProvider(next)];
+      if (current && current.type === next.type) {
+        //The data array must be edited in place. As chart.js adds listeneres to it.
+        current.data.splice(next.data.length);
+        next.data.forEach((point, pid) => {
+          current.data[pid] = next.data[pid];
+        });
+        const { data, ...otherProps } = next;
+        //Merge properties. Notice weekness here if a property is removed
```
Comment (benmccann) [inline src/index.js]: weekness -> weakness

File: src/index.js [hunk 4]
```diff
@@ -160,36 +160,35 @@ class ChartComponent extends React.Component {
...
-      const currentDatasetKey = this.props.datasetKeyProvider(currentDatasets[idx]);
-      if (nextDatasetKeys.indexOf(currentDatasetKey) === -1) {
-        // deleted series
-        currentDatasets.splice(idx, 1);
+    const currentDatasetsIndexed = keyBy(
+      currentDatasets,
+      this.props.datasetKeyProvider
+    );
+
+    //We can safely replace the dataset array, as long as we retain the _meta property
```
Comment (benmccann) [inline src/index.js]: should put a space after each `//` for consistency with existing code

**Comment by sjurba** [comment]
I read the first comment and thought: "spell checking comments.. A bit nitpicking maybe.." Then I read the second comment and was embarrassed. Then I read the third one. Sorry about that, and thank you for spell checking my comments. Fixed and amended.

**Comment by benmccann** [comment]
Thanks! I'll let @jerairrest review this PR since he's more familiar with the code. I just noticed those small issues

**Comment by jerairrest** [comment]
Looks good to me!

**Pull request merged**